aptos-move/framework/aptos-framework/sources/multisig_account.move

```diff
@@ -0,0 +1,1476 @@
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+
+        assert!(
+            table::contains(&multisig_account_resource.transactions, sequence_number),
+            error::not_found(ETRANSACTION_NOT_FOUND),
+        );
+        let transaction = table::borrow_mut(&mut multisig_account_resource.transactions, sequence_number);
+        let votes = &mut transaction.votes;
+        let owner_addr = address_of(owner);
+
+        if (simple_map::contains_key(votes, &owner_addr)) {
+            *simple_map::borrow_mut(votes, &owner_addr) = approved;
+        } else {
+            simple_map::add(votes, owner_addr, approved);
+        };
+
+        emit_event(
+            &mut multisig_account_resource.vote_events,
+            VoteEvent {
+                owner: owner_addr,
+                sequence_number,
+                approved,
+            }
+        );
+    }
+
+    /// Remove the next transaction if it has sufficient owner rejections.
+    public entry fun execute_rejected_transaction(
+        owner: &signer,
+        multisig_account: address,
+    ) acquires MultisigAccount {
+        assert_multisig_account_exists(multisig_account);
+        let multisig_account_resource = borrow_global_mut<MultisigAccount>(multisig_account);
+        assert_is_owner(owner, multisig_account_resource);
+        let sequence_number = multisig_account_resource.last_executed_sequence_number + 1;
+        assert!(
+            table::contains(&multisig_account_resource.transactions, sequence_number),
+            error::not_found(ETRANSACTION_NOT_FOUND),
+        );
+        // Delete the transaction to keep storage efficient. Off-chain components can reconstruct the transaction and
+        // from events.
+        let transaction = table::remove(&mut multisig_account_resource.transactions, sequence_number);
+        let (_, num_rejections) = num_approvals_and_rejections(&multisig_account_resource.owners, &transaction);
+        assert!(
+            num_rejections >= multisig_account_resource.num_signatures_required,
+            error::invalid_state(ENOT_ENOUGH_REJECTIONS),
+        );
+
+        multisig_account_resource.last_executed_sequence_number = sequence_number;
+        emit_event(
+            &mut multisig_account_resource.execute_rejected_transaction_events,
+            ExecuteRejectedTransactionEvent {
+                sequence_number,
+                num_rejections,
+                executor: address_of(owner),
+            }
+        );
+    }
+
+    /// Called by the VM as part of transaction prologue, which is invoked during mempool transaction validation and as
+    /// the first step of transaction execution.
+    ///
+    /// Transaction payload is optional if it's already stored on chain for the transaction.
+    fun validate_multisig_transaction(
```

Comment:
I'd call this multisig_prologue to align with other prologues, also it seems missing a few checks like balance >= gas, chain_id etc.

I'm also hesitant about checking the hash in prologue, it's expensive and we don't charge gas for prologue

Reply:
The prologue done for multisig already includes the standard tx prologue called - it'll check gas, etc. of the sender (which is not the same as the multisig account address).

If we check the hash in the prologue, the transaction would fail earlier (as mempool validation invokes this while still in the submit transaction request's scope). If we eliminate this check and do this later in execution, transactions would get dropped and failures can be less transparent to the user.

Reply:
I don't see where the common prologue is called? I only see the check_gas and this function

Reply:
I was referring to the gas param checks in aptos_vm::check_gas but you're right that we should execute the same transaction validation done in script prologue for multisig txs as well. This ensures that checks such as sufficient balance to cover gas, etc. are still done. I added the script_prologue call in multisig prologue path (see aptos_vm::run_prologue_with_payload). Will add some api tests to make sure these validations are done.